nexus/reconfigurator/planning/src/blueprint_builder/builder/datasets_editor.rs

```diff
@@ -0,0 +1,358 @@
+                BlueprintDatasetDisposition::Expunged => continue,
+                // Potentially expungeable
+                BlueprintDatasetDisposition::InService => (),
+            }
+            if expunge_if(&*dataset) {
+                dataset.disposition = BlueprintDatasetDisposition::Expunged;
+                num_expunged += 1;
+                self.changed = true;
+            }
+        }
+
+        num_expunged
+    }
+
+    pub fn ensure_debug_dataset(
+        &mut self,
+        zpool: ZpoolName,
+    ) -> (DatasetUuid, Ensure) {
+        const DEBUG_QUOTA_SIZE_GB: u32 = 100;
+
+        let address = None;
+        let quota = Some(ByteCount::from_gibibytes_u32(DEBUG_QUOTA_SIZE_GB));
+        let reservation = None;
+
+        self.ensure_dataset(
+            DatasetName::new(zpool, DatasetKind::Debug),
+            address,
+            quota,
+            reservation,
+            CompressionAlgorithm::GzipN { level: GzipLevel::new::<9>() },
+        )
+    }
+
+    pub fn ensure_zone_root_dataset(
+        &mut self,
+        zpool: ZpoolName,
+    ) -> (DatasetUuid, Ensure) {
+        let address = None;
+        let quota = None;
+        let reservation = None;
+
+        self.ensure_dataset(
+            DatasetName::new(zpool, DatasetKind::TransientZoneRoot),
+            address,
+            quota,
+            reservation,
+            CompressionAlgorithm::Off,
+        )
+    }
+
+    /// Ensures a dataset exists on this sled.
+    ///
+    /// - If the dataset exists in the blueprint already, use it.
+    /// - Otherwise, if the dataset exists in the database, re-use the UUID, but
+    ///   add it to the blueprint.
+    /// - Otherwse, create a new dataset in the blueprint, which will propagate
+    ///   to the database during execution.
+    pub fn ensure_dataset(
+        &mut self,
+        dataset: DatasetName,
+        address: Option<SocketAddrV6>,
+        quota: Option<ByteCount>,
+        reservation: Option<ByteCount>,
+        compression: CompressionAlgorithm,
+    ) -> (DatasetUuid, Ensure) {
+        let zpool_id = dataset.pool().id();
+        let kind = dataset.dataset();
+
+        let make_config = |id: DatasetUuid| BlueprintDatasetConfig {
+            disposition: BlueprintDatasetDisposition::InService,
+            id,
+            pool: dataset.pool().clone(),
+            kind: kind.clone(),
+            address,
+            quota,
+            reservation,
+            compression,
+        };
+
+        // Is this dataset already in the blueprint? If so, update it if it's
+        // changed.
+        if let Some(existing_id) = self
+            .blueprint_dataset_ids
+            .get(&zpool_id)
+            .and_then(|kind_to_id| kind_to_id.get(kind))
+        {
+            let old_config = self.config.datasets.get_mut(existing_id).expect(
+                "internal inconsistency: \
+                 entry in blueprint_dataset_ids but not current",
+            );
```

Comment:
I noticed we return some internal inconsistencies as errors, but this one causes a panic. Is there a reason for that?

Reply:
Mm I think I overused the term "internal inconsistency". In this case, something _in this module_ is wrong: we built `self.blueprint_dataset_ids` from the contents of `self.config.datasets`, so the panic here is "we built it wrong (which should never happen)". `BlueprintDatasetsEditError::MultipleDatasetsOfKind`, on the other hand, is an inconsistency problem with incoming data: either the parent blueprint or the database has a zpool with more than one dataset of the same kind on it (which shouldn't be possible in the grand scheme of things, but we shouldn't panic if given bad inputs).

I'll change the error message on `MultipleDatasetsOfKind` to indicate that it's a data consistency / input error.

Reply:
Added a comment and reworded the error in [e4d4f9b](https://github.com/oxidecomputer/omicron/pull/7104/commits/e4d4f9b73cfb3ccb3305e78c72b63ae9d5081c8a)